diff --git a/src/simunpack.c b/src/simunpack.c
index 3c5cb8c..c4d7817 100644
--- a/src/simunpack.c
+++ b/src/simunpack.c
@@ -35,7 +35,7 @@ simunpack(unsigned char *cpack, g2int *idrstmpl, g2int ndpts,
     dscale = int_power(10.0, -idrstmpl[2]);
     nbits = idrstmpl[3];
 
-    if (!(ifld = calloc(ndpts, sizeof(g2int))))
+    if (!(ifld = calloc(ndpts ? ndpts : 1, sizeof(g2int))))
     {
         fprintf(stderr, "Could not allocate space in simunpack.\n  "
 		"Data field NOT upacked.\n");
