diff --git a/src/dec_jpeg2000.c b/src/dec_jpeg2000.c
index 90333e2..8ba4adb 100644
--- a/src/dec_jpeg2000.c
+++ b/src/dec_jpeg2000.c
@@ -62,7 +62,7 @@ int dec_jpeg2000(char *injpc,g2int bufsize,g2int *outfld)
 //
 //     Decode JPEG200 codestream into jas_image_t structure.
 //
-    image=jpc_decode(jpcstream,opts);
+    image=jas_image_decode(jpcstream, -1, 0);
     if ( image == 0 ) {
         printf(" jpc_decode return\n");
         return -3;
diff --git a/src/enc_jpeg2000.c b/src/enc_jpeg2000.c
index 67079c7..2a2f603 100644
--- a/src/enc_jpeg2000.c
+++ b/src/enc_jpeg2000.c
@@ -58,6 +58,7 @@ enc_jpeg2000(unsigned char *cin, g2int width, g2int height, g2int nbits,
 	     g2int jpclen)
 {
     int ier,rwcnt;
+    int fmt;
     jas_image_t image;
     jas_stream_t *jpcstream,*istream;
     jas_image_cmpt_t cmpt,*pcmpt;
@@ -145,6 +146,8 @@ enc_jpeg2000(unsigned char *cin, g2int width, g2int height, g2int nbits,
 //
 //     Encode image.
 //
+    fmt = jas_image_strtofmt("jpc");
+    ier=jas_image_encode(&image,jpcstream,fmt,opts);
     ier=jpc_encode(&image,jpcstream,opts);
     if ( ier != 0 ) {
         printf(" jpc_encode return = %d \n",ier);
