diff --git a/src/enc_jpeg2000.c b/src/enc_jpeg2000.c
index 67079c7..97ea571 100644
--- a/src/enc_jpeg2000.c
+++ b/src/enc_jpeg2000.c
@@ -108,7 +108,9 @@ enc_jpeg2000(unsigned char *cin, g2int width, g2int height, g2int nbits,
     image.clrspc_=JAS_CLRSPC_SGRAY;         /* grayscale Image */
     image.cmprof_=0;
 #endif
-//    image.inmem_=1;
+#if JAS_VERSION_MAJOR == 1
+    image.inmem_=1;
+#endif
 
     cmpt.tlx_=0;
     cmpt.tly_=0;
