diff --git a/test/test_examples.py b/test/test_examples.py
index 5b8e913..d0f7e8e 100644
--- a/test/test_examples.py
+++ b/test/test_examples.py
@@ -31,6 +31,7 @@ def test_amenity_list(capsys):
     assert 835 == len(output)
     assert '1.535245 42.556681 parking' == output[0].strip()
     assert '1.570729 42.529562 parking         Aparcament Comunal' == output[-1].strip()
+test_amenity_list.__test__ = False
 
 
 def test_road_length(capsys):
