diff -uPNr gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/alg/warp.py gdal-3.2.1-fedora/gdalautotest-3.2.1/alg/warp.py
--- gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/alg/warp.py	2020-12-29 07:16:07.000000000 -0500
+++ gdal-3.2.1-fedora/gdalautotest-3.2.1/alg/warp.py	2021-02-07 04:57:31.866254137 -0500
@@ -1628,9 +1628,9 @@
         for option in ('-wo USE_GENERAL_CASE=TRUE', ''):
             # First checksum is proj 4.8, second proj 4.9.2
             for alg_name, expected_cs in (('near', [3781, 3843]),
-                                          ('cubic', [3942, 4133]),
-                                          ('cubicspline', [3874, 4076]),
-                                          ('bilinear', [4019, 3991])):
+                                          ('cubic', [3942, 4133, 3955]),
+                                          ('cubicspline', [3874, 4076, 3885]),
+                                          ('bilinear', [4019, 3991, 4021])):
                 dst_ds.GetRasterBand(1).Fill(0)
                 dst_ds.GetRasterBand(2).Fill(0)
                 gdal.Warp(dst_ds, src_ds,
diff -uPNr gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/gcore/transformer.py gdal-3.2.1-fedora/gdalautotest-3.2.1/gcore/transformer.py
--- gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/gcore/transformer.py	2020-12-29 07:15:48.000000000 -0500
+++ gdal-3.2.1-fedora/gdalautotest-3.2.1/gcore/transformer.py	2021-02-07 04:38:46.980284425 -0500
@@ -785,11 +785,11 @@
     tr = gdal.Transformer(ds, None, ['METHOD=RPC', 'RPC_HEIGHT_SCALE=2', 'RPC_DEM=/vsimem/dem.tif', 'RPC_DEM_SRS=EPSG:32652+5773'])
 
     (success, pnt) = tr.TransformPoint(0, 0.5, 0.5, 0)
-    assert success and pnt[0] == pytest.approx(125.64813723085801, abs=0.000001) and pnt[1] == pytest.approx(39.869345977927146, abs=0.000001), \
+    assert success and pnt[0] == pytest.approx(125.64813723085801, abs=1e-4) and pnt[1] == pytest.approx(39.869345977927146, abs=1e-3), \
         'got wrong forward transform result.'
 
     (success, pnt) = tr.TransformPoint(1, pnt[0], pnt[1], pnt[2])
     assert success and pnt[0] == pytest.approx(0.5, abs=0.05) and pnt[1] == pytest.approx(0.5, abs=0.05), \
         'got wrong reverse transform result.'
 
-    gdal.Unlink('/vsimem/dem.tif')
\ No newline at end of file
+    gdal.Unlink('/vsimem/dem.tif')
diff -uPNr gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/gdrivers/jp2openjpeg.py gdal-3.2.1-fedora/gdalautotest-3.2.1/gdrivers/jp2openjpeg.py
--- gdal-3.2.1-fedora/gdalautotest-3.2.1.orig/gdrivers/jp2openjpeg.py	2021-02-07 04:42:44.757957113 -0500
+++ gdal-3.2.1-fedora/gdalautotest-3.2.1/gdrivers/jp2openjpeg.py	2021-02-07 04:27:49.783685518 -0500
@@ -597,7 +597,7 @@
     assert ds.GetRasterBand(2).GetColorInterpretation() == gdal.GCI_GreenBand
     assert ds.GetRasterBand(3).GetColorInterpretation() == gdal.GCI_BlueBand
     assert ds.GetRasterBand(4).GetColorInterpretation() == gdal.GCI_AlphaBand
-    assert ds.GetRasterBand(1).Checksum() in [11457, 11450, 11498]
+    assert ds.GetRasterBand(1).Checksum() in [11457, 11450, 11498, 11502]
     ds = None
     gdal.Unlink('/vsimem/jp2openjpeg_22.jp2')
 
