diff -rupN --no-dereference gdal-3.2.1-fedora/apps/GNUmakefile gdal-3.2.1-fedora-new/apps/GNUmakefile
--- gdal-3.2.1-fedora/apps/GNUmakefile	2020-12-29 13:15:46.000000000 +0100
+++ gdal-3.2.1-fedora-new/apps/GNUmakefile	2021-01-04 21:49:36.444247832 +0100
@@ -233,6 +233,7 @@ gdal-config-inst:	gdal-config.in ../GDAL
 
 install: default
 	for f in $(BIN_LIST) ; do $(INSTALL) $$f $(DESTDIR)$(INST_BIN) ; done
+	for f in $(BIN_LIST) ; do $(INSTALL) .libs/$$f $(DESTDIR)$(INST_BIN) ; done
 	$(INSTALL_DATA) gdal_utils.h $(DESTDIR)$(INST_INCLUDE)
 	$(INSTALL) gdal-config-inst $(DESTDIR)$(INST_BIN)/gdal-config
 
