commit 2a0da8ad40250ee9f01155e0c894b2fa4171f26a
Author: Tom Hughes <tom@compton.nu>
Date:   Mon Mar 2 13:38:13 2020 +0000

    Update projection tests for changes in proj

diff --git a/test/unit/projection/proj_transform.cpp b/test/unit/projection/proj_transform.cpp
index cbaa4e13d..84419c00a 100644
--- a/test/unit/projection/proj_transform.cpp
+++ b/test/unit/projection/proj_transform.cpp
@@ -172,8 +172,8 @@ SECTION("Test proj antimeridian bbox")
     //        274000 7173000 # top-most
     //  END
     //
-    const mapnik::box2d<double> normal(148.7667597489, -60.1222810241,
-                                       159.9548489296, -24.9771195155);
+    const mapnik::box2d<double> normal(148.7639922894, -60.1221489796,
+                                       159.9548476477, -24.9771194497);
 
     {
         // checks for not being snapped (ie. not antimeridian)
