From: Jochen Topf <jochen@topf.org>
Date: Tue, 24 Aug 2021 11:25:15 +0200
Subject: Run ANALYZE on middle tables only in create mode

Fixes #1556
---
 src/middle-pgsql.cpp | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/src/middle-pgsql.cpp b/src/middle-pgsql.cpp
index 12fc036f..7ee7b2da 100644
--- a/src/middle-pgsql.cpp
+++ b/src/middle-pgsql.cpp
@@ -563,7 +563,7 @@ void middle_pgsql_t::relation_delete(osmid_t osm_id)
 void middle_pgsql_t::after_nodes()
 {
     m_db_copy.sync();
-    if (m_options->flat_node_file.empty()) {
+    if (!m_options->append && m_options->flat_node_file.empty()) {
         auto const &table = m_tables.nodes();
         analyze_table(m_db_connection, table.schema(), table.name());
     }
@@ -572,15 +572,19 @@ void middle_pgsql_t::after_nodes()
 void middle_pgsql_t::after_ways()
 {
     m_db_copy.sync();
-    auto const &table = m_tables.ways();
-    analyze_table(m_db_connection, table.schema(), table.name());
+    if (!m_options->append) {
+        auto const &table = m_tables.ways();
+        analyze_table(m_db_connection, table.schema(), table.name());
+    }
 }
 
 void middle_pgsql_t::after_relations()
 {
     m_db_copy.sync();
-    auto const &table = m_tables.relations();
-    analyze_table(m_db_connection, table.schema(), table.name());
+    if (!m_options->append) {
+        auto const &table = m_tables.relations();
+        analyze_table(m_db_connection, table.schema(), table.name());
+    }
 
     // release the copy thread and its database connection
     m_copy_thread->finish();
-- 
2.30.2

