diff --git a/src/osmdbt-catchup.cpp b/src/osmdbt-catchup.cpp
index 599a381..2c76b00 100644
--- a/src/osmdbt-catchup.cpp
+++ b/src/osmdbt-catchup.cpp
@@ -24,9 +24,9 @@ static lsn_type get_lsn(Config const &config)
     lsn_type lsn;
 
     boost::filesystem::path p{config.log_dir()};
-    for (auto const &file : boost::filesystem::directory_iterator(p)) {
-        if (file.path().extension() == ".log") {
-            std::string const fn = file.path().filename().string();
+    for (boost::filesystem::directory_iterator file(p); file != boost::filesystem::directory_iterator{}; ++file) {
+        if (file->path().extension() == ".log") {
+            std::string const fn = file->path().filename().string();
             std::cmatch m;
             bool const has_match = std::regex_match(fn.c_str(), m, re);
             if (has_match && m.size() == 2) {
diff --git a/src/osmdbt-create-diff.cpp b/src/osmdbt-create-diff.cpp
index 258d1ae..e3c4294 100644
--- a/src/osmdbt-create-diff.cpp
+++ b/src/osmdbt-create-diff.cpp
@@ -543,9 +543,9 @@ bool app(osmium::VerboseOutput &vout, Config const &config,
         vout << "No log files on command line. Looking for log files in log "
                 "directory...\n";
         boost::filesystem::path p{config.log_dir()};
-        for (auto const &file : boost::filesystem::directory_iterator(p)) {
-            if (file.path().extension() == ".log") {
-                log_files.push_back(file.path().filename().string());
+        for (boost::filesystem::directory_iterator file(p); file != boost::filesystem::directory_iterator{}; ++file) {
+            if (file->path().extension() == ".log") {
+                log_files.push_back(file->path().filename().string());
             }
         }
     }
