diff --git a/regress/core/tests.mk b/regress/core/tests.mk
index 517b56a..5f55041 100644
--- a/regress/core/tests.mk
+++ b/regress/core/tests.mk
@@ -145,8 +145,7 @@ endif
 ifeq ($(shell expr "$(POSTGIS_GEOS_VERSION)" ">=" 39),1)
 	# GEOS-3.0 adds stable maximuminscribedcircle implementation
 	TESTS += \
-		$(topsrcdir)/regress/core/geos39 \
-		$(topsrcdir)/regress/core/fixedoverlay
+		$(topsrcdir)/regress/core/geos39
 endif
 
 ifeq ($(INTERRUPTTESTS),yes)
diff --git a/regress/core/tests.mk.in b/regress/core/tests.mk.in
index db0e5c2..a343d60 100644
--- a/regress/core/tests.mk.in
+++ b/regress/core/tests.mk.in
@@ -145,8 +145,7 @@ endif
 ifeq ($(shell expr "$(POSTGIS_GEOS_VERSION)" ">=" 39),1)
 	# GEOS-3.0 adds stable maximuminscribedcircle implementation
 	TESTS += \
-		$(topsrcdir)/regress/core/geos39 \
-		$(topsrcdir)/regress/core/fixedoverlay
+		$(topsrcdir)/regress/core/geos39
 endif
 
 ifeq ($(INTERRUPTTESTS),yes)
