diff -rupN proj-7.2.0/CMakeLists.txt proj-7.2.0-new/CMakeLists.txt
--- proj-7.2.0/CMakeLists.txt	2020-10-28 12:22:26.000000000 +0100
+++ proj-7.2.0-new/CMakeLists.txt	2020-11-05 07:27:39.486134481 +0100
@@ -277,3 +277,18 @@ add_subdirectory(cmake)
 if(BUILD_TESTING)
   add_subdirectory(test)
 endif()
+
+################################################################################
+# pkgconfig file
+################################################################################
+SET(prefix ${CMAKE_INSTALL_PREFIX})
+SET(exec_prefix ${CMAKE_INSTALL_PREFIX})
+SET(libdir ${CMAKE_INSTALL_PREFIX}/${LIBDIR})
+SET(includedir ${CMAKE_INSTALL_PREFIX}/${INCLUDEDIR})
+SET(datarootdir ${CMAKE_INSTALL_PREFIX}/${DATADIR})
+SET(VERSION ${PROJ_VERSION})
+SET(SQLITE3_LIBS ${SQLITE3_LIBRARY})
+SET(TIFF_LIBS ${TIFF_LIBRARY})
+SET(CURL_LIBS ${CURL_LIBRARY})
+CONFIGURE_FILE(proj.pc.in proj.pc @ONLY)
+INSTALL(FILES ${CMAKE_CURRENT_BINARY_DIR}/proj.pc DESTINATION ${LIBDIR}/pkgconfig)
