diff --git a/gbfile.cc b/gbfile.cc
index 0884798..7203280 100644
--- a/gbfile.cc
+++ b/gbfile.cc
@@ -306,7 +306,6 @@ stdapi_seek(gbfile* self, int32_t offset, int whence)
 static gbsize_t
 stdapi_read(void* buf, const gbsize_t size, const gbsize_t members, gbfile* self)
 {
-  int errno;
   gbsize_t result = fread(buf, size, members, self->handle.std);
 
   if ((result < members) && (errno = ferror(self->handle.std))) {
