commit 92e9b684944399ad4f8683b37450f89bdd2902b3
Author: David Hummel <6109326+hummeltech@users.noreply.github.com>
Date:   Mon May 11 20:20:58 2020 -0700

    Another proj 6.x test results patch

diff --git a/test/python_tests/projection_test.py b/test/python_tests/projection_test.py
index 316632928..591949ece 100644
--- a/test/python_tests/projection_test.py
+++ b/test/python_tests/projection_test.py
@@ -139,8 +139,8 @@ def test_proj_antimeridian_bbox():
     prj_trans_rev = mapnik.ProjTransform(prjGeog, prjProj)
 
     # bad = mapnik.Box2d(-177.31453250437079, -62.33374815225163, 178.02778363316355, -24.584597490955804)
-    better = mapnik.Box2d(-180.0, -62.33374815225163,
-                          180.0, -24.584597490955804)
+    better = mapnik.Box2d(-180.0, -62.3337481528,
+                          180.0, -24.5845959277)
 
     buffered_query_ext = mapnik.Box2d(274000, 3087000, 3327000, 7173000)
     fwd_ext = prj_trans_fwd.forward(buffered_query_ext, PROJ_ENVELOPE_POINTS)
