diff --git a/spec/ruby/library/socket/socket/bind_spec.rb b/spec/ruby/library/socket/socket/bind_spec.rb
index 0349df8..1fb3e90 100644
--- a/spec/ruby/library/socket/socket/bind_spec.rb
+++ b/spec/ruby/library/socket/socket/bind_spec.rb
@@ -30,15 +30,6 @@ describe "Socket#bind on SOCK_DGRAM socket" do
     sockaddr1 = Socket.pack_sockaddr_in(0, "4.3.2.1")
     -> { @sock.bind(sockaddr1) }.should raise_error(Errno::EADDRNOTAVAIL)
   end
-
-  platform_is_not :windows, :cygwin do
-    as_user do
-      it "raises Errno::EACCES when the current user does not have permission to bind" do
-        sockaddr1 = Socket.pack_sockaddr_in(1, "127.0.0.1")
-        -> { @sock.bind(sockaddr1) }.should raise_error(Errno::EACCES)
-      end
-    end
-  end
 end
 
 describe "Socket#bind on SOCK_STREAM socket" do
@@ -71,15 +62,6 @@ describe "Socket#bind on SOCK_STREAM socket" do
     sockaddr1 = Socket.pack_sockaddr_in(0, "4.3.2.1")
     -> { @sock.bind(sockaddr1) }.should raise_error(Errno::EADDRNOTAVAIL)
   end
-
-  platform_is_not :windows, :cygwin do
-    as_user do
-      it "raises Errno::EACCES when the current user does not have permission to bind" do
-        sockaddr1 = Socket.pack_sockaddr_in(1, "127.0.0.1")
-        -> { @sock.bind(sockaddr1) }.should raise_error(Errno::EACCES)
-      end
-    end
-  end
 end
 
 describe 'Socket#bind' do
@@ -110,16 +92,6 @@ describe 'Socket#bind' do
 
         -> { @socket.bind(sockaddr1) }.should raise_error(Errno::EADDRNOTAVAIL)
       end
-
-      platform_is_not :windows do
-        as_user do
-          it 'raises Errno::EACCES when the user is not allowed to bind to the port' do
-            sockaddr1 = Socket.pack_sockaddr_in(1, ip_address)
-
-            -> { @socket.bind(sockaddr1) }.should raise_error(Errno::EACCES)
-          end
-       end
-      end
     end
 
     describe 'using an Addrinfo' do
